mitmproxy/dns.py

```diff
@@ -0,0 +1,869 @@
+        try:
+            addrinfos = socket.getaddrinfo(host=self.name, port=0, family=family)
+        except socket.gaierror as e:
+            if e.errno == socket.EAI_NODATA:
+                raise ResolveError(ResponseCode.NXDOMAIN)
+            else:
+                # NOTE might fail on Windows for IPv6 queries:
+                # https://stackoverflow.com/questions/66755681/getaddrinfo-c-on-windows-not-handling-ipv6-correctly-returning-error-code-1
+                raise ResolveError(ResponseCode.SERVFAIL)
+        for addrinfo in addrinfos:
+            _, _, _, _, addr = addrinfo
+            yield ResourceRecord(
+                name=self.name,
+                type=self.type,
+                class_=self.class_,
+                ttl=ResourceRecord.DEFAULT_TTL,
+                data=ip(addr[0]).packed,
+            )
+
+    def _resolve_by_addr(
+        self,
+        suffix: str,
+        ip: Callable[[List[str]], Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]
+    ) -> ResourceRecord:
+        try:
+            addr = ip(self.name[:-len(suffix)].split(".")[::-1])
+        except ValueError:
+            raise ResolveError(ResponseCode.FORMERR)
+        try:
+            name, _, _ = socket.gethostbyaddr(str(addr))
+        except socket.herror:
+            raise ResolveError(ResponseCode.NXDOMAIN)
+        except socket.gaierror:
+            raise ResolveError(ResponseCode.SERVFAIL)
+        return ResourceRecord(
+            name=self.name,
+            type=self.type,
+            class_=self.class_,
+            ttl=ResourceRecord.DEFAULT_TTL,
+            data=ResourceRecord.pack_domain_name(name),
+        )
+
+    def resolve(self) -> Generator[ResourceRecord, None, None]:
+        """Resolve the question into resource record(s), throwing ResolveError if an error condition occurs."""
+
+        if self.class_ is not Class.IN:
+            raise ResolveError(ResponseCode.NOTIMP)
+        if self.type is Type.A:
+            yield from self._resolve_by_name(socket.AddressFamily.AF_INET, ipaddress.IPv4Address)
+        elif self.type is Type.AAAA:
+            yield from self._resolve_by_name(socket.AddressFamily.AF_INET6, ipaddress.IPv6Address)
+        elif self.type is Type.PTR:
+            name_lower = self.name.lower()
+            if name_lower.endswith(Question.IP4_PTR_SUFFIX):
+                yield self._resolve_by_addr(Question.IP4_PTR_SUFFIX, lambda x: ipaddress.IPv4Address(".".join(x)))
+            elif name_lower.endswith(Question.IP6_PTR_SUFFIX):
+                yield self._resolve_by_addr(Question.IP6_PTR_SUFFIX, lambda x: ipaddress.IPv6Address(bytes.fromhex("".join(x))))
+            else:
+                raise ResolveError(ResponseCode.FORMERR)
+        else:
+            raise ResolveError(ResponseCode.NOTIMP)
+
+
+@dataclass
+class ResourceRecord(BypassInitStateObject):
+    # since preferable every query should go through mitmproxy, keep the TTL as low as possible
+    DEFAULT_TTL = 1
```

Comment:
If `dns_mode=forward:...` is used, upstream's TTL will be returned to the client (if no `dns_response` hook decides to change it). If `simple` mode is used, we don't really get a TTL from `getaddrinfo`, so I felt it should be as low as possible by default, in case someone wants to get a flow for every time a client tries to resolve an address. Totally open to different default suggestions, though.

Reply:
Let's assume a minute, see https://textslashplain.com/2022/03/31/chromiums-dns-cache/

Reply:
Let's assume a minute, see https://textslashplain.com/2022/03/31/chromiums-dns-cache/

Reply:
Changed to 1 minute in f535045e69207402f224f8f28c355caec84040e6.